nx/lib/nx.ex

```diff
@@ -6408,15 +6408,158 @@ defmodule Nx do
+  and batch axes.
+
+
+  The contracting axes must be dot-product compatible and the batch dimensions must
+  always have the same number of elements.
+
+  ## Examples
+
+  ### Dot product between two batched tensors
+
+      iex> u = Nx.tensor([[[1]], [[2]]])
+      iex> v = Nx.tensor([[[3]], [[4]]])
+      iex> Nx.dot(u, [2], [0], v, [2], [0])
+      #Nx.Tensor<
+        s64[2][1][1]
+        [
+          [
+            [3]
+          ],
+          [
+            [8]
+          ]
+        ]
+      >
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [2], [0], v, [1], [0])
+      #Nx.Tensor<
+        s64[2][1][1]
+        [
+          [
+            [6]
+          ],
+          [
+            [16]
+          ]
+        ]
+      >
+
+  ### Error cases
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [2], [0], v, [1], [])
+      ** (ArgumentError) right tensor must be batched if left tensor is batched
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [2], [], v, [1], [0])
+      ** (ArgumentError) left tensor must be batched if right tensor is batched
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [2], [1], v, [1], [0])
+      ** (ArgumentError) invalid dot batch axis for the left tensor - only batch axis 0 is supported, but got [1]
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [2], [0], v, [1], [1])
+      ** (ArgumentError) invalid dot batch axis for the right tensor - only batch axis 0 is supported, but got [1]
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [0], [0], v, [1], [0])
+      ** (ArgumentError) dot batch axes [0] for the left tensor cannot be in the contract axes [0]
+
+      iex> u = Nx.tensor([[[1, 1]], [[2, 2]]])
+      iex> v = Nx.tensor([[[3], [3]], [[4], [4]]])
+      iex> Nx.dot(u, [2], [0], v, [0], [0])
+      ** (ArgumentError) dot batch axes [0] for the right tensor cannot be in the contract axes [0]
+  """
+  @doc type: :ndim
+  def dot(t1, contract_axes1, batch_axes1, t2, contract_axes2, batch_axes2) do
+    t1 = to_tensor(t1)
+    t2 = to_tensor(t2)
+
     output_type = binary_type(t1, t2)
-    %T{shape: s1, names: names1} = t1 = to_tensor(t1)
-    %T{shape: s2, names: names2} = t2 = to_tensor(t2)
-    axes1 = Nx.Shape.normalize_axes(s1, axes1, names1)
-    axes2 = Nx.Shape.normalize_axes(s2, axes2, names2)
-    {output_shape, output_names} = Nx.Shape.zip_reduce(s1, axes1, names1, s2, axes2, names2)
+
+    {s1, names1, c1, b1, s2, names2, c2, b2} =
+      Nx.Shape.dot(t1, contract_axes1, batch_axes1, t2, contract_axes2, batch_axes2)
```

Comment:
Btw, I would expect a function named `Nx.Shape.dot` to return the output shape of `dot`. It can also perform validation, but generally speaking it computes the shape. You can also make it compute the `{output_shape, batch_each_shape}` so you can reuse it elsewhere.

Reply:
This makes sense. Perhaps some of the validation will need to be duplicated if the intermediate values aren't to be returned. I'll look into maybe changing the function signature to accept the normalized axes instead. WDYT?

Reply:
Yes, typically axes normalization is done outside of Nx.Shape.